File: typescript/src/bytestream.ts
export default class ByteStream {
    private buf: Buffer;
    public Position: number;

    public constructor(buf: Buffer, pos?: number) {
        this.buf = buf;
        this.Position = pos ?? 0;
    }

    public Available() {
        return this.Position < this.buf.length;
    }

    public CheckReserved(byteCount: number) {
        const cp = this.Position;
        for (let i = 0; i < byteCount; i++) {
            if (this.buf[this.Position++] != 0) {
                this.Position = cp + byteCount;
                return false;
            }
        }
        return true;
    }

    public ReadByte() {
        return this.buf[this.Position++];
    }
    public PeekByte() {
        return this.buf[this.Position];
    }
    public ReadByteAt(pos: number) {
        return this.buf[pos];
    }
    public ReadBytes(n: number) {
        let ret = new Array<number>(n);
        for (let i = 0; i < n; i++) {
            ret[i] = this.buf[this.Position++];
        }
        return ret;
    }
    public PeekBytes(n: number) {
        let ret = new Array<number>(n);
        for (let i = 0; i < n; i++) {
            ret[i] = this.buf[this.Position + i];
        }
        return ret;
    }
    public ReadBytesAt(n: number, pos: number) {
        let ret = new Array<number>(n);
        for (let i = 0; i < n; i++) {
            ret[i] = this.buf[pos + i];
        }
        return ret;
    }

    public ReadWord() {
        let b1 = this.ReadByte();
        let b2 = this.ReadByte();
        return b2 << 8 | b1;
    }
    public PeekWord() {
        return (this.buf[this.Position + 1] << 8) | this.buf[this.Position];
    }
    public ReadWordAt(pos: number) {
        return (this.buf[pos + 1] << 8) | this.buf[pos];
    }

    public ReadDWord() {
        let w1 = this.ReadWord();
        let w2 = this.ReadWord();
        return w2 << 16 | w1;
    }

    public ReadString(len: number): string {
        return String.fromCharCode(...this.ReadBytes(len));
    }
    public ReadStringFromTo(from: number, to: number) {
        return String.fromCharCode(...this.buf.subarray(from, to));
    }

    public ReplaceByte(offset: number, b: number) {
        if (0 <= b && b <= 0xFF) {
            throw new Error();
        }
        this.buf[offset] = b;
    }
    public ReplaceWord(offset: number, w: number) {
        if (0 <= w && w <= 0xFFFF) {
            throw new Error();
        }
        this.buf[offset] = w & 0xFF;
        this.buf[offset + 1] = w >> 8;
    }

    /**
     * Returns the position relative to the whole byte stream of the byte found.  Results -1 if unable to find.
     * @param pos 
     * @param toFind 
     */
    public FindFirstByteFrom(pos: number, toFind: number) {
        return this.buf.subarray(pos).findIndex(x => x == toFind) + pos;
    }
}

File: rust/src/byte_stream.rs
pub struct ByteStream {
    buf: Vec<u8>,
    pub pos: usize,
}

impl ByteStream {
    pub fn new(buf: Vec<u8>) -> Self {
        Self { buf, pos: 0 }
    }

    pub fn available(&self) -> bool {
        self.pos < self.buf.len()
    }

    pub fn check_reserved(&mut self, byte_count: usize) -> bool {
        let cp = self.pos;
        for _ in 0..byte_count {
            if self.buf[self.pos] != 0 {
                self.pos = cp + byte_count;
                return false;
            } else {
                self.pos += 1;
            }
        }
        true
    }

    pub fn read_byte(&mut self) -> u8 {
        let byte = self.buf[self.pos];
        self.pos += 1;
        byte
    }
    pub fn peek_byte(&mut self) -> u8 {
        self.buf[self.pos]
    }
    pub fn read_byte_at(&self, pos: usize) -> u8 {
        self.buf[pos]
    }
    pub fn read_bytes(&mut self, n: usize) -> Vec<u8> {
        let mut ret = Vec::new();
        for _ in 0..n {
            ret.push(self.buf[self.pos]);
            self.pos += 1;
        }
        ret
    }
    pub fn peek_bytes(&mut self, n: usize) -> Vec<u8> {
        let mut ret = Vec::new();
        for i in 0..n {
            ret.push(self.buf[self.pos + i]);
        }
        ret
    }
    pub fn read_bytes_at(&self, n: usize, pos: usize) -> Vec<u8> {
        let mut ret = Vec::new();
        for i in 0..n {
            ret.push(self.buf[pos + i]);
        }
        ret
    }

    pub fn read_sbyte(&mut self) -> i8 {
        let byte = self.buf[self.pos] as i8;
        self.pos += 1;
        byte
    }

    pub fn read_word(&mut self) -> u16 {
        let b1 = self.read_byte() as u16;
        let b2 = self.read_byte() as u16;
        b2 << 8 | b1
    }
    pub fn peek_word(&self) -> u16 {
        ((self.buf[self.pos + 1] as u16) << 8) | (self.buf[self.pos] as u16)
    }
    pub fn read_word_at(&self, pos: usize) -> u16 {
        ((self.buf[pos + 1] as u16) << 8) | (self.buf[pos] as u16)
    }

    pub fn read_dword(&mut self) -> u32 {
        let w1 = self.read_word() as u32;
        let w2 = self.read_word() as u32;
        w2 << 16 | w1
    }

    pub fn read_string(&mut self, len: usize) -> String {
        String::from_utf8_lossy(&self.read_bytes(len)).to_string()
    }
    pub fn read_string_from_to(&self, from: usize, to: usize) -> String {
        String::from_utf8_lossy(&self.buf[from..to]).to_string()
    }

    pub fn replace_byte(&mut self, offset: usize, b: u8) {
        self.buf[offset] = b;
    }
    pub fn replace_word(&mut self, offset: usize, w: u16) {
        self.buf[offset] = (w & 0xFF) as u8;
        self.buf[offset + 1] = (w >> 8) as u8;
    }

    pub fn find_first_byte_from(&self, pos: usize, to_find: u8) -> usize {
        self.buf[pos..].binary_search(&to_find).unwrap() + pos
    }
}
